src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,692 @@
+        self.units.push(unit);
+        index
+    }
+
+    /// Compute a topological sort of the unit graph, only including libraries
+    /// that should be compiled by mir-json.
+    fn sequence_libs(&self) -> Vec<MirJsonLib> {
+        let mut result = Vec::new();
+        // Cache for the result of the visit function.
+        let mut visited: Vec<Option<Option<&CrateName>>> =
+            vec![None; self.units.len()];
+
+        /// Process the unit at index `i`. Return its crate name if it should be
+        /// included as a dependency, or [None] otherwise.
+        fn visit<'a>(
+            units: &'a [CustomUnit],
+            result: &mut Vec<MirJsonLib>,
+            visited: &mut [Option<Option<&'a CrateName>>],
+            i: usize,
+        ) -> Option<&'a CrateName> {
+            if let Some(r) = visited[i] {
+                return r;
+            }
+            let r = match &units[i].target {
+                CustomTarget::TargetLib(lib) => {
+                    // Visit all dependencies and add this unit to result.
+                    let mut lib_deps = Vec::new();
+                    for dep in &units[i].dependencies {
+                        if let Some(real_crate_name) =
+                            visit(units, result, visited, dep.index)
+                        {
+                            lib_deps.push((
+                                dep.extern_crate_name.clone(),
+                                real_crate_name.clone(),
+                            ));
+                        }
+                    }
+                    result.push(MirJsonLib {
+                        target: lib.clone(),
+                        dependencies: lib_deps,
+                    });
+                    Some(&lib.crate_name)
+                }
+                CustomTarget::HostDep => {
+                    // Do not visit dependencies and do not add this unit to
+                    // result.
+                    None
+                }
+                CustomTarget::FinalBin => {
+                    // Visit all dependencies but do not add this unit to
+                    // result.
+                    for dep in &units[i].dependencies {
+                        visit(units, result, visited, dep.index);
+                    }
+                    None
+                }
+            };
+            visited[i] = Some(r);
+            r
+        }
+
+        for root in &self.roots {
+            visit(&self.units, &mut result, &mut visited, *root);
+        }
+
+        result
+    }
+}
+
+/// A representation of a command that can be easily formatted as a shell
+/// command or turned into an actual [Command].
+struct CmdInvocation {
+    program: String,
+    args: Vec<String>,
+    env: Vec<(String, String)>,
+}
+
+impl CmdInvocation {
+    fn as_command(&self) -> Command {
+        let mut cmd = Command::new(&self.program);
+        cmd.args(&self.args)
+            .envs(self.env.iter().map(|(k, v)| (k, v)));
```

Comment:
For my own Rust learning: can you explain why the purpose of the `.map(|(k, v)| (k, v))` call is? At first glance, that looks like a no-op, but perhaps there is a subtle type change going on there.

Reply:
There is indeed a subtle type change. `Vec<(String, String)>::iter` returns an `Iterator<Item = &(String, String)>`, but `Command::envs` expects an `IntoIterator<Item = (K, V)>` where `K, V: AsRef<OsStr>`. So the `map` is converting `&(String, String)` to `(&String, &String)`. But pattern-matching on a `&(String, String)` automatically binds each element to `&String`, so we just put them back together in a pair again.

Reply:
Might be clearer to make this explicit: `|&(ref k, ref v)| (k, v)`